tests/integration/test_boltdb_v11_to_tsdb_v12_migration.py

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This test module tests loki upgrade to validate schema migration from v11/boltdb to v12/tsdb.
+
+Test Scenarios:
+1. Deploy from Charmhub with revision 140 using v11/boltdb, then upgrade to local charm with v12/tsdb.
+2. Deploy locally with v12/tsdb, then upgrade locally to validate configuration persistence.
+"""
+
+import datetime
+import logging
+from pathlib import Path
+
+import pytest
+import yaml
+from helpers import is_loki_up, loki_config
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+V11_APP_NAME = f'v11-{METADATA["name"]}'
+V12_APP_NAME = f'v12-{METADATA["name"]}'
+
+resources = {
+    "loki-image": METADATA["resources"]["loki-image"]["upstream-source"],
+    "node-exporter-image": METADATA["resources"]["node-exporter-image"]["upstream-source"],
+}
+
+
+async def test_setup_env(ops_test: OpsTest):
+    await ops_test.model.set_config({"logging-config": "<root>=WARNING; unit=DEBUG"})
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_from_charmhub_v11_and_upgrade_to_v12_locally(ops_test: OpsTest, loki_charm):
+    """Deploy from Charmhub (v11 schema) and upgrade to v12 locally."""
+    await deploy_charm_from_charmhub_v11(ops_test, V11_APP_NAME)
+    await upgrade_charm_with_local_charm_v12(ops_test, V11_APP_NAME, loki_charm)
+    await verify_upgrade_success(ops_test, V11_APP_NAME, False)
+
+    # Here we upgrade again to ensure config is persisted and won't be overwritten with any weird values
+    await upgrade_charm_with_local_charm_v12(ops_test, V11_APP_NAME, loki_charm)
+    await verify_upgrade_success(ops_test, V11_APP_NAME, False)
```

Comment:
Are these test going to be valid once this PR is merged and a brand new `edge` version released?

Reply:
The test is locked on deploying rev 140 from charmhub, so it should yes.